manifests/type.pp

```diff
@@ -1,29 +1,35 @@
 define collectd::type (
-  $target,
-  $ds_type,
-  $ds = $title,
-  $min = 'U',
-  $max = 'U',
-  $ds_name = 'value',
-) {
-  validate_string($ds_name)
-  $upper_ds_type = upcase($ds_type)
-
-  validate_re($upper_ds_type,
-              ['^ABSOLUTE$', '^COUNTER$', '^DERIVE$', '^GAUGE$'])
+  String                                                      $target,
+  String                                                      $ds = $name,
+  # BC compatible ....
+  Optional[Enum['ABSOLUTE', 'COUNTER', 'DERIVE', 'GAUGE']]    $ds_type = undef,
+  Variant[Numeric, Enum['U']]                                 $min = 'U',
+  Variant[Numeric, Enum['U']]                                 $max = 'U',
+  String                                                      $ds_name = 'value',
+  # BC compatible .... ending
 
-  if $min != 'U' {
-    validate_numeric($min)
-  }
-
-  if $max != 'U' {
-    validate_numeric($max)
+  Array[Struct[{
+    min     => Variant[Numeric, Enum['U']],
+    max     => Variant[Numeric, Enum['U']],
+    ds_type => Enum['ABSOLUTE', 'COUNTER', 'DERIVE', 'GAUGE'],
+    ds_name => String,
+  }]]                                                         $types = [],
```

Comment:
ah, now I understand the layout. I'm not aware of any current best practices for the layout, we will keep this as is.